README
Sudoku version 0.01
=========================

Extensible Sudoku Puzzle and solver.  See pod documentation for more.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SCRIPT

You can watch the solver work by running 

./sudoku.pl puzzles/test1.puz

or test2.puz, etc.  Be warned, solving Hex puzzles is quite slow right now.  sudoku.pl is not installed by make install.

DEPENDENCIES

This module requires these other modules and libraries:

  Moose
  

COPYRIGHT AND LICENCE

Copyright (C) 2012 by Josh Burroughs <josh@qhool.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Sudoku::Board.

0.01  Mon Apr 23 22:59:20 2012
	- original version; created by h2xs 1.23 with options
		-AX -b 5.8.0 Sudoku::Board



SudokuBoard.pm
package SudokuBoard;

use SudokuSquare;
use Data::Dumper;
use Storable qw(dclone);
use Moose;
use Scalar::Util qw(weaken);

=head1 NAME

SudokuBoard

=head1 SYNOPSIS

  use SudokuBoard;
  $board = SudokuBoard->new();

  $board->set_puzzle($txt);
  $board->display();
  $board->solve();
  $ans = $board->is_solved();

  $num = $board->get_num_unset();

=head1 CONSTRUCTOR

=over

=item new()

Creates new SudokuBoard.  Generally no arguments are given.

=back

=head1 METHODS

=head2 width()
=head2 height()
=head2 block_width()
=head2 block_height()

These return fixed values for the width & height of a puzzle, and the width & height of the sub-blocks within it.

=cut 

my @dims = qw(width height block_width block_height);
for my $i (0..3) {
  #this breaks Moose's 'fourth wall' a bit, but I didn't want to type this stuff 4 times
  has $dims[$i] => 
    ( is => 'ro',
      isa => 'Int',
      default => sub { my $self = shift; 
		       return $self->_dimensions()->[$i]; 
		     }
    );
}

=head2 square(X,Y)

Returns the SudokuSquare object at the given coordinates in the puzzle.

=cut

sub square {
  my $self = shift;
  my $idx = join(':',@_);
  return $self->squares()->{$idx};
}

=head2 all_squares()

Returns list of all SudokuSquare objects in the puzzle.  Useful for iterating over all squares.

=cut

sub all_squares {
  my $self = shift;
  return values( %{$self->squares()} );
}

=head2 set_puzzle( PUZZLE_STRING )

Sets the values in the puzzle based on the string given, which should have one line per row of the puzzle, with each character on each line representing one square.  A space is used for each open square.

=cut

sub set_puzzle {
  my $self = shift;
  my $puz_str = shift;
  my @lines = split( /\n/m, $puz_str );
  if( @lines != $self->height() ) {
    die "wrong number of lines!";
  }
  for my $i (0..$#lines) {
    chomp($lines[$i]);
    my @v = split( //, $lines[$i] );
     for my $j (0..$#v) {
      #print "$i,$j = $v[$j]\n";
       if( $v[$j] !~ /\s/ ) {
	 $self->square($i,$j)->from_str($v[$j]);
       }
    }
  }
}

=head2 display()

Prints an ascii-art representation of the puzzle to STDOUT.

=cut

sub display {
  my $self = shift;
  my $disp_type = 'VAL';
  if( @_ > 0 ) {
    $disp_type = shift;
  }
  my $h_bar = "-" x ($self->width()*2 + 1);
  print "$h_bar\n";
  for my $i (0..$self->height()-1) {
    print "|";
    for my $j (0..$self->width()-1) {
      if( $disp_type eq 'VAL' ) {
	print $self->square($i,$j)->to_str();
      } elsif( $disp_type eq 'OPT' ) {
	if( defined( $self->square($i,$j)->value() ) ) {
	  print ' ';
	} else {
	  print $self->square($i,$j)->get_num_possible();
	}
      } else {
	print ' ';
      }
      if( ($j+1) % $self->block_width() == 0 ) {
	print '|';
      } else {
	print ' ';
      }
    }
    if( ($i+1) % $self->block_height() == 0 ) {
      print "\n$h_bar\n";
    } else {
      print "\n";
    }
  }
}

=head2 get_num_unset()

Returns the number of squares without a definite value().

=cut

sub get_num_unset {
  my $self = shift;
  my $num_unset = 0;
  map { $num_unset++ } grep {not defined} map { $_->value() } $self->all_squares();
  return $num_unset;
}

=head2 is_solved()

Returns a true value if all squares have a definite value.

=cut

sub is_solved {
  my $self = shift;
  return (0 == $self->get_num_unset());
}

=head2 solve(DISPLAY)

Attempts to solve the puzzle.  Returns true if successful.  First attempts deterministic methods, then tries guessing a single value, followed by re-application of deterministic methods, backtracking when the puzzle is found to be unsolvable.

=cut

sub solve {
  my ($self,$display,$depth) = @_;
  $depth ||= 1;
  if( $self->solve_det($display) ) {
    return 1;
  } else {
    #guess values, then try to continue solving
    my @uncertain_squares = 
      map { $_->[1] }
	sort { $a->[0] <=> $b->[0] }
	  map { [$_->get_num_possible(), $_] }
	    grep { not defined $_->value() } $self->all_squares();
    my $orig_squares = $self->squares();
    for my $unc (@uncertain_squares) {
      for my $v ($unc->possibilities()) {
	#clone the board, so we can revert later if needed
	my $squares_dupe = dclone($orig_squares);
	$self->squares( dclone($orig_squares) );
	#get the clone of the current square in question
	my $sq = $self->square( @{$unc->index()} );
	print " " . ("." x $depth) . "Trying $v at (" . join(",",@{$sq->index()}) . ")\n";
	$sq->value($v);
	my $sol = 
	  eval {
	    $self->solve($display,$depth+1);
	  };
	if( $@ and $@ !~ /UNSOLVABLE/ ) {
	  die $@;
	} elsif( $sol ) {
	  return 1;
	}
      }
    }
    #nothing yielded a solution:
    return 0;
  }
}

=head1 SUBCLASSING

To create a subclass, you should override at least _square_class() and _dimensions().  See HexdokuBoard.pm for an example.

=head1 INTERNALS

The following functions are used internally by SudokuBoard.

=head2 square_class()

=head2 _square_class()

Return the name of the class to use for creating squares in the puzzle.  When subclassing, override _square_class() -- it is a Moose builder sub for square_class().

=cut

has 'square_class' =>
( is => 'ro',
  isa => 'ClassName',
  builder => '_square_class'
);

sub _square_class {
  return 'SudokuSquare';
}

=head2 squares()

Returns the hash ref of all squares in the puzzle.

=cut

has 'squares' =>
( is => 'rw',
  isa => 'HashRef[SudokuSquare]',
  builder => '_build_squares',
);

=head2 _dimensions()

Returns an array ref containing the width, height, width of sub-blocks, height of sub-blocks.  Used by the width() method, etc.

=cut

sub _dimensions {
  my $self = shift;
  return [9,9,3,3];
}

=head2 _square_indexes()

Returns a list of the indexes of all squares in the puzzle, based on the values in _dimensions().  Override for 3D or otherwise non-rectangular puzzles.

=cut

sub _square_indexes {
  my $self = shift;
  my @idx;
  for my $i (0..$self->width()-1) {
    for my $j (0..$self->height()-1) {
      push @idx, [$i,$j]
    }
  }
  return @idx;
}

=head2 _coset_indexes()

Return a list of array refs, one for each co-set (group of cells which must contain all possible values).  In normal Sudoku, this is the rows, columns, and three-by-three blocks.  Cosets are build based on _dimensions().

=cut

sub _coset_indexes {
  my $self = shift;
  my @cosets;
  #row sets
  for my $i (0..$self->height()-1) {
    push @cosets, [map { [$i,$_] } (0..$self->width()-1) ];
  }
  #column sets
  for my $i (0..$self->width()-1) {
    push @cosets, [map { [$_,$i] } (0..$self->height()-1) ];
  }

  for( my $w = 0; $w < $self->width(); $w += $self->block_width() ) {
    for( my $h = 0; $h < $self->height(); $h += $self->block_height() ) {
      my @block;
      for my $i (0..$self->block_width()-1) {
	for my $j (0..$self->block_height()-1) {
	  push @block, [$w+$i,$h+$j];
	}
      }
      push @cosets, \@block;
    }
  }
  return @cosets;
}

=head2 _build_squares()

Builder method for squares(); creates SudokuSquare objects to populate the board.  Behavior is controlled by _square_indexes() and _coset_indexes().

=cut

sub _build_squares {
  my $self = shift;
  my %squares;

  for my $index ($self->_square_indexes()) {
    my $idx = join(":",@$index);
    my $sq = $self->square_class()->new( index => $index );
    $squares{$idx} = $sq;

    if( $squares{$idx} ne $sq ) {
      die "Mismatch!";
    }
  }

  #do all the cosets:
  for my $cos ($self->_coset_indexes()) {
    my @coset = map { my $ref = $squares{$_}; weaken($ref); $ref } map { join(':',@$_); } @$cos;
    for my $sq (@coset) {
      $sq->add_coset( @coset );
    }
  }
  return \%squares;
}

=head2 solve_det( DISPLAY ) 

Tries deterministic solution methods.  If DISPLAY is true, calls display() after each pass.  Returns true value if puzzle is solved.  Dies with "UNSOLVABLE" if puzzle is not solvable.

=cut

sub solve_det {
  my $self = shift;
  my $display = 0;
  if( @_ ) {
    $display = shift;
  }
  my $num_iter = 0;
  while( 1 ) {
    my $num_improved = 0;
    for my $sq ($self->all_squares()) {
      if( $sq->compare_to_cosets() ) {
	$num_improved++;
      }
    }
    if( $display ) {
      $self->display();
    }
    $num_iter++;
    if( $num_improved == 0 ) {
      if( $self->is_solved() ) {
	print "Solved!\n";
	return 1;
      } else {
	return 0;
      }
    }
  }
}



1;

#end


SudokuSquare.pm
package SudokuSquare;

use Moose;

=head1 NAME

SudokuSquare

=head1 SYNOPSIS

  use SudokuSquare;
  $sq = SudokuSquare->new(\@index);

  @possible = $sq->possibilities();
  $val = $sq->value();
  $sq->value(7); #set the value

  $sq->compare_to_cosets();

=head1 DESCRIPTION

The SudokuSquare class is meant to be used by SudokuBoard.  It keeps track of the value or possible values of the square, as well as the cosets (my name for the regions in a sudoku puzzle where each value must appear exactly once, and can perform inferences about its value based on the 

=head1 SUBCLASSING

To make a sub-class of SudokuSquare, you need to override all_possibilities(), and possibly from_str() and to_str() if the display form of values are anything besides numbers (e.g. for hexadecimal sudoku).  See HexdokuSquare.pm in this package for an example.

=head1 CONSTRUCTOR

=over 4

=item new( index => $idx );

Create a new SudokuSquare.  $idx should be an arrayref of coordinates for this square within the puzzle (it's an array to allow for more than 2 dimensions).

=back

=head1 METHODS

=head2 value( [ VALUE ] )

Get or set the value of the square.  Return undef for an un-set square.

=cut
has 'value' =>
(
 is => 'rw',
 isa => 'Int',
 trigger => \&_set_value
);

=head2 possible()

Returns a hash-ref of possible values.  Keys are the possibilities, and values are 1 for allowable possibilities, 0 for ruled-out possibilities.

=cut

has 'possible' =>
( is => 'ro',
  isa => 'HashRef[Bool]',
  builder => '_build_possible'
);

=head2 cosets()

Returns array ref of cosets.  Use add_coset() to modify cosets of the Square.

=cut

has 'cosets' =>
( is => 'ro',
  isa => 'ArrayRef[ArrayRef[SudokuSquare]]',
  default => sub { [] }
);

=head2 index( INDEX1, INDEX2, ... )

Get or set the index of the square.

=cut

has 'index' =>
( is => 'rw',
  isa => 'ArrayRef',
);

=head2 all_possibilities()

Return list of all possible values for Sudoku squares.  Should be integers only.

=cut

sub all_possibilities {
  return (1..9);
}

=head2 from_str( STRING )

Set the value for the square from $str.  Does nothing if $str consists of whitespace.  Used by SudokuBoard::set_puzzle() to construct a puzzle from a string.

=cut

sub from_str {
  my $self = shift;
  my $str = shift;
  if( $str !~ /^\s*$/ ) {
    return $self->value($str);
  }
}

=head2 to_str()

Returns a textual representation of the Square.  Used by SudokuSquare::display().

=cut

sub to_str {
  my $self = shift;
  return $self->value() || ' ';
}


sub _build_possible {
  my $self = shift;
  return { map { $_ => 1 } $self->all_possibilities() };
}

sub add_coset {
  my $self = shift;
  push @{$self->cosets()}, [@_];
}

sub _unset_possibilities {
  my $self = shift;
  for my $k (keys(%{$self->possible()})) {
    $self->possible()->{$k} = 0;
  }
}

sub _set_value {
  my $self = shift;
  $self->_unset_possibilities();
  $self->possible()->{$self->value()};
}

sub set_possibilities {
  my $self = shift;
  my @vals = @_;
  $self->_unset_possibilities();
  for my $v (@vals) {
    $self->possible()->{$v} = 1;
  }
}

sub get_num_possible {
  my $self = shift;
  return 0 + $self->possibilities();
}

sub possibilities {
  my $self = shift;
  return grep { $self->possible()->{$_} } $self->all_possibilities();
}

=head2 deduce()

Tries to solve for the value of this square by eliminating as possibilities any known values from the squares in the cosets of this square.  If the value of this square is set, returns it; otherwise undef.

=cut

sub deduce {
  my $self = shift;
  if( defined($self->value()) ) {
    return $self->value();
  }
  for my $coset (@{$self->cosets()}) {
    map { $self->possible()->{$_} = 0 }
      grep { defined } 
	map { $_->value() } @{$coset};
  }
  my @possible = $self->possibilities();
  if( @possible == 1 ) {
    return $self->value($possible[0]);
  } elsif( @possible == 0 ) {
    die "UNSOLVABLE";
  }
  return undef;
}

sub sort_unique {
  my @t = sort(@_);
  my @r;
  for my $t (@t) {
    if( @r == 0 or $t != $r[$#r] ) {
      push @r, $t;
    }
  }
  return @r;
}

sub minus {
  my @a = sort_unique @{shift(@_)};
  my @b = sort_unique @{shift(@_)};
  my @c;
  while( @a and @b ) {
    if( $a[0] < $b[0] ) {
      unshift @c, shift(@a);
    } elsif( $a[0] == $b[0] ) {
      shift @a; shift @b;
    } else {
      shift @b;
    }
  }
  return (@c,@a);
}

sub intersect {
  my @a = sort_unique @{shift(@_)};
  my @b = sort_unique @{shift(@_)};
  my @c;
  while( @a and @b ) {
    if( $a[0] == $b[0] ) {
      push @c, shift(@a); shift @b;
    } elsif( $a[0] < $b[0] ) {
      shift @a;
    } else {
      shift @b;
    }
  }
  return @c;
}

=head2 induce()

Tries to solve for the value of this square by looking for any values which are not possible for another square in one of the cosets of this square.  Returns the value of the square if it is set, undef otherwise.

=cut

sub induce {
  my $self = shift;
  if( defined($self->value()) ) {
    return $self->value();
  }
  my @vals = $self->possibilities();
  for my $coset (@{$self->cosets()}) {
    my @vals = 
      intersect( \@vals,
		 [ minus( [$self->all_possibilities()],
			  [ map { $_->possibilities() } @{$coset} ] ) ] );
  }
  if( @vals == 1 ) {
    print "Induced: ", join(",",@vals), "\n";
    $self->value( $vals[0] );
  } elsif( @vals == 0 ) {
    die "UNSOLVABLE";
  } else {
    $self->set_possibilities( @vals );
  }
  return $self->value();
}

=head2 compare_to_cosets()

Calls deduce() and induce().  If the value of the square is set during the call, or if the number of possible values is reduced, returns 1; otherwise returns 0.  SudokuBoard uses this return value to tell when the deterministic solver has 'settled' -- if the calls to compare_to_cosets for all squares return 0, that means nothing has changed, and the solution is stuck.

=cut

sub compare_to_cosets {
  my $self = shift;
  if( defined $self->value() ) {
    return 0;
  }
  my $possible_before = $self->get_num_possible();
  $self->deduce;
  $self->induce;
  if( defined $self->value() or $possible_before > $self->get_num_possible() ) {
    return 1;
  } else {
    return 0;
  }
}

=head1 TODO

Improve deterministic solution with X-Wings, etc.

=cut

1;
#end



lib/Sudoku/Square.pm
package Sudoku::Square;

use 5.008000;
use strict;
use warnings;

use Moose;

our $VERSION = '0.1';

=head1 NAME

Sudoku::Square

=head1 SYNOPSIS

  use Sudoku::Square;
  $sq = Sudoku::Square->new(\@index);

  @possible = $sq->possibilities();
  $val = $sq->value();
  $sq->value(7); #set the value

  $sq->compare_to_cosets();

=head1 DESCRIPTION

The Sudoku::Square class is meant to be used by Sudoku::Board.  It keeps track of the value or possible values of the square, as well as the cosets (my name for the regions in a sudoku puzzle where each value must appear exactly once, and can perform inferences about its value based on the 

=head1 SUBCLASSING

To make a sub-class of Sudoku::Square, you need to override all_possibilities(), and possibly from_str() and to_str() if the display form of values are anything besides numbers (e.g. for hexadecimal sudoku).  See Sudoku::HexSquare.pm in this package for an example.

=head1 CONSTRUCTOR

=over 4

=item new( index => $idx );

Create a new Sudoku::Square.  $idx should be an arrayref of coordinates for this square within the puzzle (it's an array to allow for more than 2 dimensions).

=back

=head1 METHODS

=head2 value( [ VALUE ] )

Get or set the value of the square.  Return undef for an un-set square.

=cut
has 'value' =>
(
 is => 'rw',
 isa => 'Int',
 trigger => \&_set_value
);

=head2 possible()

Returns a hash-ref of possible values.  Keys are the possibilities, and values are 1 for allowable possibilities, 0 for ruled-out possibilities.

=cut

has 'possible' =>
( is => 'ro',
  isa => 'HashRef[Bool]',
  builder => '_build_possible'
);

=head2 cosets()

Returns array ref of cosets.  Use add_coset() to modify cosets of the Square.

=cut

has 'cosets' =>
( is => 'ro',
  isa => 'ArrayRef[ArrayRef[Sudoku::Square]]',
  default => sub { [] }
);

=head2 index( INDEX1, INDEX2, ... )

Get or set the index of the square.

=cut

has 'index' =>
( is => 'rw',
  isa => 'ArrayRef',
);

=head2 all_possibilities()

Return list of all possible values for Sudoku squares.  Should be integers only.

=cut

sub all_possibilities {
  return (1..9);
}

=head2 from_str( STRING )

Set the value for the square from $str.  Does nothing if $str consists of whitespace.  Used by Sudoku::Board::set_puzzle() to construct a puzzle from a string.

=cut

sub from_str {
  my $self = shift;
  my $str = shift;
  if( $str !~ /^\s*$/ ) {
    return $self->value($str);
  }
}

=head2 to_str()

Returns a textual representation of the Square.  Used by Sudoku::Square::display().

=cut

sub to_str {
  my $self = shift;
  return $self->value() || ' ';
}


sub _build_possible {
  my $self = shift;
  return { map { $_ => 1 } $self->all_possibilities() };
}

sub add_coset {
  my $self = shift;
  push @{$self->cosets()}, [@_];
}

sub _unset_possibilities {
  my $self = shift;
  for my $k (keys(%{$self->possible()})) {
    $self->possible()->{$k} = 0;
  }
}

sub _set_value {
  my $self = shift;
  $self->_unset_possibilities();
  $self->possible()->{$self->value()};
}

sub set_possibilities {
  my $self = shift;
  my @vals = @_;
  $self->_unset_possibilities();
  for my $v (@vals) {
    $self->possible()->{$v} = 1;
  }
}

sub get_num_possible {
  my $self = shift;
  return 0 + $self->possibilities();
}

sub possibilities {
  my $self = shift;
  return grep { $self->possible()->{$_} } $self->all_possibilities();
}

=head2 deduce()

Tries to solve for the value of this square by eliminating as possibilities any known values from the squares in the cosets of this square.  If the value of this square is set, returns it; otherwise undef.

=cut

sub deduce {
  my $self = shift;
  if( defined($self->value()) ) {
    return $self->value();
  }
  for my $coset (@{$self->cosets()}) {
    map { $self->possible()->{$_} = 0 }
      grep { defined } 
	map { $_->value() } @{$coset};
  }
  my @possible = $self->possibilities();
  if( @possible == 1 ) {
    return $self->value($possible[0]);
  } elsif( @possible == 0 ) {
    die "UNSOLVABLE";
  }
  return undef;
}

sub sort_unique {
  my @t = sort(@_);
  my @r;
  for my $t (@t) {
    if( @r == 0 or $t != $r[$#r] ) {
      push @r, $t;
    }
  }
  return @r;
}

sub minus {
  my @a = sort_unique @{shift(@_)};
  my @b = sort_unique @{shift(@_)};
  my @c;
  while( @a and @b ) {
    if( $a[0] < $b[0] ) {
      unshift @c, shift(@a);
    } elsif( $a[0] == $b[0] ) {
      shift @a; shift @b;
    } else {
      shift @b;
    }
  }
  return (@c,@a);
}

sub intersect {
  my @a = sort_unique @{shift(@_)};
  my @b = sort_unique @{shift(@_)};
  my @c;
  while( @a and @b ) {
    if( $a[0] == $b[0] ) {
      push @c, shift(@a); shift @b;
    } elsif( $a[0] < $b[0] ) {
      shift @a;
    } else {
      shift @b;
    }
  }
  return @c;
}

=head2 induce()

Tries to solve for the value of this square by looking for any values which are not possible for another square in one of the cosets of this square.  Returns the value of the square if it is set, undef otherwise.

=cut

sub induce {
  my $self = shift;
  if( defined($self->value()) ) {
    return $self->value();
  }
  my @vals = $self->possibilities();
  for my $coset (@{$self->cosets()}) {
    my @vals = 
      intersect( \@vals,
		 [ minus( [$self->all_possibilities()],
			  [ map { $_->possibilities() } @{$coset} ] ) ] );
  }
  if( @vals == 1 ) {
    print "Induced: ", join(",",@vals), "\n";
    $self->value( $vals[0] );
  } elsif( @vals == 0 ) {
    die "UNSOLVABLE";
  } else {
    $self->set_possibilities( @vals );
  }
  return $self->value();
}

=head2 compare_to_cosets()

Calls deduce() and induce().  If the value of the square is set during the call, or if the number of possible values is reduced, returns 1; otherwise returns 0.  Sudoku::Board uses this return value to tell when the deterministic solver has 'settled' -- if the calls to compare_to_cosets for all squares return 0, that means nothing has changed, and the solution is stuck.

=cut

sub compare_to_cosets {
  my $self = shift;
  if( defined $self->value() ) {
    return 0;
  }
  my $possible_before = $self->get_num_possible();
  $self->deduce;
  $self->induce;
  if( defined $self->value() or $possible_before > $self->get_num_possible() ) {
    return 1;
  } else {
    return 0;
  }
}

=head1 TODO

Improve deterministic solution with X-Wings, etc.

=cut

1;
__END__



lib/Sudoku/Board.pm
package Sudoku::Board;

use 5.008000;
use strict;
use warnings;

use Sudoku::Square;
use Data::Dumper;
use Storable qw(dclone);
use Moose;
use Scalar::Util qw(weaken);

our $VERSION = '0.1';

=head1 NAME

SudokuBoard

=head1 SYNOPSIS

  use Sudoku::Board;
  $board = Sudoku::Board->new();

  $board->set_puzzle($txt);
  $board->display();
  $board->solve();
  $ans = $board->is_solved();

  $num = $board->get_num_unset();

=head1 CONSTRUCTOR

=over

=item new()

Creates new Sudoku::Board.  Generally no arguments are given.

=back

=head1 METHODS

=head2 width()
=head2 height()
=head2 block_width()
=head2 block_height()

These return fixed values for the width & height of a puzzle, and the width & height of the sub-blocks within it.

=cut 

my @dims = qw(width height block_width block_height);
for my $i (0..3) {
  #this breaks Moose's 'fourth wall' a bit, but I didn't want to type this stuff 4 times
  has $dims[$i] => 
    ( is => 'ro',
      isa => 'Int',
      default => sub { my $self = shift; 
		       return $self->_dimensions()->[$i]; 
		     }
    );
}

=head2 square(X,Y)

Returns the Sudoku::Square object at the given coordinates in the puzzle.

=cut

sub square {
  my $self = shift;
  my $idx = join(':',@_);
  return $self->squares()->{$idx};
}

=head2 all_squares()

Returns list of all Sudoku::Square objects in the puzzle.  Useful for iterating over all squares.

=cut

sub all_squares {
  my $self = shift;
  return values( %{$self->squares()} );
}

=head2 set_puzzle( PUZZLE_STRING )

Sets the values in the puzzle based on the string given, which should have one line per row of the puzzle, with each character on each line representing one square.  A space is used for each open square.

=cut

sub set_puzzle {
  my $self = shift;
  my $puz_str = shift;
  my @lines = split( /\n/m, $puz_str );
  if( @lines != $self->height() ) {
    die "wrong number of lines!";
  }
  for my $i (0..$#lines) {
    chomp($lines[$i]);
    my @v = split( //, $lines[$i] );
     for my $j (0..$#v) {
      #print "$i,$j = $v[$j]\n";
       if( $v[$j] !~ /\s/ ) {
	 $self->square($i,$j)->from_str($v[$j]);
       }
    }
  }
  return 1;
}

=head2 display()

Prints an ascii-art representation of the puzzle to STDOUT.

=cut

sub display {
  my $self = shift;
  my $disp_type = 'VAL';
  if( @_ > 0 ) {
    $disp_type = shift;
  }
  my $h_bar = "-" x ($self->width()*2 + 1);
  print "$h_bar\n";
  for my $i (0..$self->height()-1) {
    print "|";
    for my $j (0..$self->width()-1) {
      if( $disp_type eq 'VAL' ) {
	print $self->square($i,$j)->to_str();
      } elsif( $disp_type eq 'OPT' ) {
	if( defined( $self->square($i,$j)->value() ) ) {
	  print ' ';
	} else {
	  print $self->square($i,$j)->get_num_possible();
	}
      } else {
	print ' ';
      }
      if( ($j+1) % $self->block_width() == 0 ) {
	print '|';
      } else {
	print ' ';
      }
    }
    if( ($i+1) % $self->block_height() == 0 ) {
      print "\n$h_bar\n";
    } else {
      print "\n";
    }
  }
}

=head2 get_num_unset()

Returns the number of squares without a definite value().

=cut

sub get_num_unset {
  my $self = shift;
  my $num_unset = 0;
  map { $num_unset++ } grep {not defined} map { $_->value() } $self->all_squares();
  return $num_unset;
}

=head2 is_solved()

Returns a true value if all squares have a definite value.

=cut

sub is_solved {
  my $self = shift;
  return (0 == $self->get_num_unset());
}

=head2 solve(DISPLAY)

Attempts to solve the puzzle.  Returns true if successful.  First attempts deterministic methods, then tries guessing a single value, followed by re-application of deterministic methods, backtracking when the puzzle is found to be unsolvable.

=cut

sub solve {
  my ($self,$display,$depth) = @_;
  $depth ||= 1;
  if( $self->solve_det($display) ) {
    return 1;
  } else {
    #guess values, then try to continue solving
    my @uncertain_squares = 
      map { $_->[1] }
	sort { $a->[0] <=> $b->[0] }
	  map { [$_->get_num_possible(), $_] }
	    grep { not defined $_->value() } $self->all_squares();
    my $orig_squares = $self->squares();
    for my $unc (@uncertain_squares) {
      for my $v ($unc->possibilities()) {
	#clone the board, so we can revert later if needed
	my $squares_dupe = dclone($orig_squares);
	$self->squares( dclone($orig_squares) );
	#get the clone of the current square in question
	my $sq = $self->square( @{$unc->index()} );
	print " " . ("." x $depth) . "Trying $v at (" . join(",",@{$sq->index()}) . ")\n";
	$sq->value($v);
	my $sol = 
	  eval {
	    $self->solve($display,$depth+1);
	  };
	if( $@ and $@ !~ /UNSOLVABLE/ ) {
	  die $@;
	} elsif( $sol ) {
	  return 1;
	}
      }
    }
    #nothing yielded a solution:
    return 0;
  }
}

=head1 SUBCLASSING

To create a subclass, you should override at least _square_class() and _dimensions().  See Sudoku::HexBoard.pm for an example.

=head1 INTERNALS

The following functions are used internally by Sudoku::Board.

=head2 square_class()

=head2 _square_class()

Return the name of the class to use for creating squares in the puzzle.  When subclassing, override _square_class() -- it is a Moose builder sub for square_class().

=cut

has 'square_class' =>
( is => 'ro',
  isa => 'ClassName',
  builder => '_square_class'
);

sub _square_class {
  return 'Sudoku::Square';
}

=head2 squares()

Returns the hash ref of all squares in the puzzle.

=cut

has 'squares' =>
( is => 'rw',
  isa => 'HashRef[Sudoku::Square]',
  builder => '_build_squares',
);

=head2 _dimensions()

Returns an array ref containing the width, height, width of sub-blocks, height of sub-blocks.  Used by the width() method, etc.

=cut

sub _dimensions {
  my $self = shift;
  return [9,9,3,3];
}

=head2 _square_indexes()

Returns a list of the indexes of all squares in the puzzle, based on the values in _dimensions().  Override for 3D or otherwise non-rectangular puzzles.

=cut

sub _square_indexes {
  my $self = shift;
  my @idx;
  for my $i (0..$self->width()-1) {
    for my $j (0..$self->height()-1) {
      push @idx, [$i,$j]
    }
  }
  return @idx;
}

=head2 _coset_indexes()

Return a list of array refs, one for each co-set (group of cells which must contain all possible values).  In normal Sudoku, this is the rows, columns, and three-by-three blocks.  Cosets are build based on _dimensions().

=cut

sub _coset_indexes {
  my $self = shift;
  my @cosets;
  #row sets
  for my $i (0..$self->height()-1) {
    push @cosets, [map { [$i,$_] } (0..$self->width()-1) ];
  }
  #column sets
  for my $i (0..$self->width()-1) {
    push @cosets, [map { [$_,$i] } (0..$self->height()-1) ];
  }

  for( my $w = 0; $w < $self->width(); $w += $self->block_width() ) {
    for( my $h = 0; $h < $self->height(); $h += $self->block_height() ) {
      my @block;
      for my $i (0..$self->block_width()-1) {
	for my $j (0..$self->block_height()-1) {
	  push @block, [$w+$i,$h+$j];
	}
      }
      push @cosets, \@block;
    }
  }
  return @cosets;
}

=head2 _build_squares()

Builder method for squares(); creates SudokuSquare objects to populate the board.  Behavior is controlled by _square_indexes() and _coset_indexes().

=cut

sub _build_squares {
  my $self = shift;
  my %squares;

  for my $index ($self->_square_indexes()) {
    my $idx = join(":",@$index);
    my $sq = $self->square_class()->new( index => $index );
    $squares{$idx} = $sq;

    if( $squares{$idx} ne $sq ) {
      die "Mismatch!";
    }
  }

  #do all the cosets:
  for my $cos ($self->_coset_indexes()) {
    my @coset = map { my $ref = $squares{$_}; weaken($ref); $ref } map { join(':',@$_); } @$cos;
    for my $sq (@coset) {
      $sq->add_coset( @coset );
    }
  }
  return \%squares;
}

=head2 solve_det( DISPLAY ) 

Tries deterministic solution methods.  If DISPLAY is true, calls display() after each pass.  Returns true value if puzzle is solved.  Dies with "UNSOLVABLE" if puzzle is not solvable.

=cut

sub solve_det {
  my $self = shift;
  my $display = 0;
  if( @_ ) {
    $display = shift;
  }
  my $num_iter = 0;
  while( 1 ) {
    my $num_improved = 0;
    for my $sq ($self->all_squares()) {
      if( $sq->compare_to_cosets() ) {
	$num_improved++;
      }
    }
    if( $display ) {
      $self->display();
    }
    $num_iter++;
    if( $num_improved == 0 ) {
      if( $self->is_solved() ) {
	print "Solved!\n";
	return 1;
      } else {
	return 0;
      }
    }
  }
}



1;
__END__
